### ISSUE TITLE
[push] Migration API to import WebPush subscriptions

### ISSUE THREAD

**Issue by jonalmeida** [mozilla/application-services#2048]
After some talks with @grigoryk regarding migration, it would be nice if there was an API where we could dump WebPush subscriptions from Fennec that can be imported into the push database OR better if we could point the schematized [`PushState`](https://searchfox.org/mozilla-esr68/source/mobile/android/base/java/org/mozilla/gecko/push/PushState.java) output file and have the native component import the subscriptions.

Once that is done, the native component could then run the equivalent of `verifyConnection` to ensure that the imported subscriptions are still valid.

┆Issue is synchronized with this [Jira Story](https://jira.mozilla.com/browse/SYNC-14)
┆Epic: Fenix Migration

**Comment by jonalmeida** [comment]
I think I mentioned this on another ticket as well, when we do the migration and do the refresh event, we need to ensure we don't DDOS the autopush servers when folks migrate.

**Comment by rfk** [comment]
Hi @jonalmeida, I don't think we were tracking this bug as part of the app-services "migration" milestone, so apologies that it fell through the cracks a little. Do you still need this ahead of the Fennec migration?

If so, do you have some existing push migration code that I could take a look at to help get my head around what would be involved here?

**Comment by rfk** [comment]
IIUC, this would look something like an `import_push_subscriptions(filePath)` method, which reads the given file to find all the push subscriptions from Fennec, sucks them into the db for the rust component, and does some server-side stuff to refresh each one.

**Comment by jonalmeida** [comment]
> IIUC, this would look something like an `import_push_subscriptions(filePath)` method, which reads the given file to find all the push subscriptions from Fennec, sucks them into the db for the rust component, and does some server-side stuff to refresh each one.

@rfk That's the most accurate description of it to date. :)

> Do you still need this ahead of the Fennec migration?

At this moment, no. This is something that can definitely come in after the most important parts of migration are in. I suspect the hardest part is just applying the data format transformations from Fennec push before adding it to the db.

**Comment by jdragojevic** [comment]
@jonalmeida  - can you be a little more specific on the whether or not this change is a must or nice-to-have?  (ie - is this something you require prior to the mid-Feb beta migration?)

**Comment by jonalmeida** [comment]
I believe it's a must for Fennec migration, but @grigoryk can confirm on that if needed.

**Comment by linacambridge** [comment]
Unfortunately, migrating this data from Fennec to Fenix is very complicated.

AIUI, the `PushState` file stores push endpoints for Web Push subscriptions. However, the mapping of push subscriptions to service worker scopes (which is what's used to dispatch the `onpushsubscriptionchange` events that @jonalmeida mentioned in https://github.com/mozilla-mobile/fenix/issues/5260#issuecomment-558823917), as well as encryption keys, are stored in an IndexedDB database, in the Gecko layer. Web Push in GeckoView doesn't use this at all. Further, we can't migrate encryption keys for subscriptions, as they're impossible to extract from IDB because the raw key bytes aren't exposed to JS.

We _can_ add a migrator to GeckoView to open the IDB database, retrieve the mapping of endpoints to service worker scopes, and notify each service worker to resubscribe. That would also require migrating the Fennec permissions database, as well as for the service workers to handle the `onpushsubscriptionchange` event. That's up to the site to handle, and many don't. ☹️

The impact of not migrating push subscriptions is that the user would have to revisit any websites they were previously subscribed to, and grant permission again.

**Comment by jdragojevic** [comment]
Agreed in a slack conversation with @linacambridge @jonalmeida @st3fan and vesta to close this as a wontfix for the migration effort.  Stefan is going to add a note to QA for Fenix to validate that they user is re-prompted by their websites to enable when they visit post-migration.
